### PR TITLE
libbpf-tools: add softirqs

### PR DESCRIPTION
**Pull request by ethercflow**
```
Tracing soft irq event time... Hit Ctrl-C to end.

16:34:08
SOFTIRQ          TOTAL_usecs
timer                    168
net_rx                  1251
tasklet                   22
sched                   1005
rcu                      212
```

Or

```
Tracing soft irq event time... Hit Ctrl-C to end.

softirq = timer
     usecs               : count    distribution
         0 -> 1          : 152      |****************************************|
         2 -> 3          : 20       |*****                                   |
         4 -> 7          : 7        |*                                       |

softirq = net_rx
     usecs               : count    distribution
         0 -> 1          : 0        |                                        |
         2 -> 3          : 0        |                                        |
         4 -> 7          : 0        |                                        |
         8 -> 15         : 1        |********************                    |
        16 -> 31         : 2        |****************************************|

softirq = tasklet
     usecs               : count    distribution
         0 -> 1          : 0        |                                        |
         2 -> 3          : 0        |                                        |
         4 -> 7          : 1        |**********                              |
         8 -> 15         : 4        |****************************************|
        16 -> 31         : 3        |******************************          |

softirq = sched
     usecs               : count    distribution
         0 -> 1          : 17       |******                                  |
         2 -> 3          : 112      |****************************************|
         4 -> 7          : 49       |*****************                       |
         8 -> 15         : 1        |                                        |
        16 -> 31         : 0        |                                        |
        32 -> 63         : 1        |                                        |
        64 -> 127        : 1        |                                        |

softirq = rcu
     usecs               : count    distribution
         0 -> 1          : 146      |****************************************|
         2 -> 3          : 12       |***                                     |
         4 -> 7          : 3        |                                        |
         8 -> 15         : 9        |**                                      |
        16 -> 31         : 3        |                                        |
```

Signed-off-by: Wenbo Zhang <ethercflow@gmail.com>

### CODE REVIEW

**Comment by ethercflow** [comment]
Hi @anakryiko, another tool PTAL when you have time, thank you. :)

File: libbpf-tools/softirqs.bpf.c [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+	tsp = bpf_map_lookup_elem(&start, &key);
+	if (!tsp || !*tsp)
+		return 0;
+
+	delta = bpf_ktime_get_ns() - *tsp;
+	if (delta < 0)
+		return 0;
+	if (!targ_ns)
+		delta /= 1000;
```
Comment (anakryiko) [inline libbpf-tools/softirqs.bpf.c:59]: BPF doesn't have signed division instruction, if this works, it's only by accident due to some Clang optimization. Please use unsigned integers when dividing

File: libbpf-tools/softirqs.c [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+static int print_hist(int fd)
+{
+	char *units = env.nanoseconds ? "nsecs" : "usecs";
+	struct hist hist, zero = {};
+	__u32 vec;
+	int err;
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++) {
+		err = bpf_map_lookup_elem(fd, &vec, &hist);
```
Comment (anakryiko) [inline libbpf-tools/softirqs.c:176]: given `NR_SOFTIRQS` is small and fixed, there is no reason to use ARRAY map, just use global array variable for this. It will be faster and simpler on both BPF and user-space side.

File: libbpf-tools/softirqs.c [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+			time(&t);
+			tm = localtime(&t);
+			strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+			printf("%-8s\n", ts);
+		}
+
+		if (!env.distributed)
+			err = print_count(bpf_map__fd(obj->maps.counts));
+		else
+			err = print_hist(bpf_map__fd(obj->maps.hists));
```
Comment (anakryiko) [inline libbpf-tools/softirqs.c:265]: similar to another PR comment, maybe pass `struct bpf_map *` as arguments?

File: libbpf-tools/softirqs.bpf.c (hunk 1 above)
Comment (ethercflow) [reply libbpf-tools/softirqs.bpf.c]: Fixed, thanks!

File: libbpf-tools/softirqs.c (hunk 2 above)
Comment (ethercflow) [reply libbpf-tools/softirqs.c]: Fixed, thanks!

File: libbpf-tools/softirqs.c (hunk 3 above)
Comment (ethercflow) [reply libbpf-tools/softirqs.c]: Sure,  as we've change bpf arrays to global values, so I change it to 

```c
static int print_hist(struct softirqs_bpf__bss *bss);
static int print_count(struct softirqs_bpf__bss *bss);
```

instead.

File: libbpf-tools/Makefile [hunk 4]
```diff
@@ -59,7 +60,7 @@ $(OUTPUT) $(OUTPUT)/libbpf:
 
 $(APPS): %: $(OUTPUT)/%.o $(LIBBPF_OBJ) $(COMMON_OBJ) | $(OUTPUT)
 	$(call msg,BINARY,$@)
-	$(Q)$(CC) $(CFLAGS) $^ -lelf -lz -o $@
+	$(Q)$(CC) $(CFLAGS) $^ -latomic -lelf -lz -o $@
```
Comment (anakryiko) [inline libbpf-tools/Makefile:63]: why is this needed? those built-ins shouldn't require extra libraries

File: libbpf-tools/softirqs.c [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+static int print_count(struct softirqs_bpf__bss *bss)
+{
+	const char *units = env.nanoseconds ? "nsecs" : "usecs";
+	__u32 vec;
+
+	printf("%-16s %6s%5s\n", "SOFTIRQ", "TOTAL_", units);
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++) {
+		if (bss->counts[vec] > 0)
```
Comment (anakryiko) [inline libbpf-tools/softirqs.c:146]: use __atomic_load_n() to read it

File: libbpf-tools/softirqs.c [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+		hist = &bss->hists[vec];
+		if (!memcmp(&zero, hist, sizeof(*hist)))
+			continue;
+		printf("softirq = %s\n", vec_names[vec]);
+		print_log2_hist(hist->slots, MAX_SLOTS, units);
+		printf("\n");
+	}
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++)
+		__atomic_store(&bss->hists[vec], &zero, __ATOMIC_RELAXED);
```
Comment (anakryiko) [inline libbpf-tools/softirqs.c:175]: how is this supposed to work? atomics are only available for a single word integers, not arbitrary structs. I guess for histograms you'll have to use map after all :(

Comment (anakryiko) [reply libbpf-tools/softirqs.c:175]: or we can just leave with races and zero out non-atomically with memset, which is probably ok as well

File: libbpf-tools/Makefile (hunk 4 above)
Comment (ethercflow) [reply libbpf-tools/Makefile:63]: I found when calling   generic version of an atomic operation, such as `__atomic_store` without `_n`, we need to link `libatomic`.

File: libbpf-tools/softirqs.c (hunk 6 above)
Comment (ethercflow) [reply libbpf-tools/softirqs.c:175]: > how is this supposed to work? atomics are only available for a single word integers, not arbitrary structs. I guess for histograms you'll have to use map after all :(

Oh I think I misunderstood the meaning of "generic version"

> Built-in Function: void __atomic_store (type *ptr, type *val, int memorder)
> This is the generic version of an atomic store. It stores the value of *val into *ptr.

Comment (ethercflow) [reply libbpf-tools/softirqs.c:175]: > how is this supposed to work? 

Um, I am a bit confused about this at the moment,  I do agree 

> atomics are only available for a single word integers, not arbitrary structs. 

But from  https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html I saw

> Built-in Function: void __atomic_store (type *ptr, type *val, int memorder)
> This is the generic version of an atomic store. It stores the value of *val into *ptr.

And from the test results, it does modify the entire structure （ `softirq -d 1` # clear  once a second）

```
softirq = timer
     usecs               : count    distribution
         0 -> 1          : 600      |****************************************|
         2 -> 3          : 39       |**                                      |
         4 -> 7          : 53       |***                                     |
         8 -> 15         : 4        |                                        |
        16 -> 31         : 13       |                                        |
        32 -> 63         : 3        |                                        |
        64 -> 127        : 2        |                                        |

softirq = timer
     usecs               : count    distribution
         0 -> 1          : 679      |****************************************|
         2 -> 3          : 30       |*                                       |
         4 -> 7          : 50       |**                                      |
         8 -> 15         : 5        |                                        |
        16 -> 31         : 5        |                                        |
        32 -> 63         : 1        |                                        |
        64 -> 127        : 3        |                                        |

softirq = timer
     usecs               : count    distribution
         0 -> 1          : 483      |****************************************|
         2 -> 3          : 30       |**                                      |
         4 -> 7          : 46       |***                                     |
         8 -> 15         : 4        |                                        |
        16 -> 31         : 4        |                                        |
        32 -> 63         : 3        |                                        |
        64 -> 127        : 3        |                                        |

softirq = timer
     usecs               : count    distribution
         0 -> 1          : 426      |****************************************|
         2 -> 3          : 22       |**                                      |
         4 -> 7          : 40       |***                                     |
         8 -> 15         : 4        |                                        |
        16 -> 31         : 2        |                                        |
        32 -> 63         : 4        |                                        |
        64 -> 127        : 1        |                                        |

softirq = timer
     usecs               : count    distribution
         0 -> 1          : 565      |****************************************|
         2 -> 3          : 41       |**                                      |
         4 -> 7          : 44       |***                                     |
         8 -> 15         : 7        |                                        |
        16 -> 31         : 9        |                                        |
        32 -> 63         : 4        |                                        |
        64 -> 127        : 1        |                                        |

softirq = timer
     usecs               : count    distribution
         0 -> 1          : 391      |****************************************|
         2 -> 3          : 22       |**                                      |
         4 -> 7          : 41       |****                                    |
         8 -> 15         : 4        |                                        |
        16 -> 31         : 1        |                                        |
        32 -> 63         : 4        |                                        |
        64 -> 127        : 1        |                                        |
```

I will determine the reason.

Comment (ethercflow) [reply libbpf-tools/softirqs.c:175]: > how is this supposed to work? atomics are only available for a single word integers, not arbitrary structs. I guess for histograms you'll have to use map after all :(

From
https://sourcegraph.com/github.com/gcc-mirror/gcc@e1336703f8220dcffdeddb5e19dd032c766fbb8f/-/blob/libatomic/config/posix/lock.c#L91
https://sourcegraph.com/github.com/gcc-mirror/gcc@e1336703f8220dcffdeddb5e19dd032c766fbb8f/-/blob/libatomic/gstore.c#L100
I see  `__atomic_store` is implement by `pthread_mutex` + `memcpy`.

I perfer 
> or we can just leave with races and zero out non-atomically with memset, which is probably ok as well

than map, thanks!

File: libbpf-tools/softirqs.c (hunk 5 above)
Comment (ethercflow) [reply libbpf-tools/softirqs.c]: Fixed, thanks!

File: libbpf-tools/softirqs.c (hunk 6 above)
Comment (ethercflow) [reply libbpf-tools/softirqs.c]: Fixed, thanks!

**Comment by ethercflow** [comment]
Hi @anakryiko! Thanks for the review, pushed fixes for all your comments. :)

File: libbpf-tools/softirqs.c [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+	printf("%-16s %6s%5s\n", "SOFTIRQ", "TOTAL_", units);
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++) {
+		count = __atomic_load_n(&bss->counts[vec], __ATOMIC_RELAXED);
+		if (count > 0)
+			printf("%-16s %11llu\n", vec_names[vec], count);
+	}
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++)
+		__atomic_store_n(&bss->counts[vec], 0, __ATOMIC_RELAXED);
```
Comment (anakryiko) [inline libbpf-tools/softirqs.c:153]: `__atomic_exchange_n` is what you want here to read value and atomically set it to zero at the same time

File: libbpf-tools/softirqs.c [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+	for (vec = 0; vec < NR_SOFTIRQS; vec++) {
+		hist = &bss->hists[vec];
+		if (!memcmp(&zero, hist, sizeof(*hist)))
+			continue;
+		printf("softirq = %s\n", vec_names[vec]);
+		print_log2_hist(hist->slots, MAX_SLOTS, units);
+		printf("\n");
+	}
+
+	memset(&bss->hists, 0, sizeof(bss->hists));
```
Comment (anakryiko) [inline libbpf-tools/softirqs.c:175]: do memset for each individual histogram maybe? it looks cleaner and behaves a bit saner, minimizing the race between reading histogram and resetting it

Comment (anakryiko) [reply libbpf-tools/softirqs.c:175]: also, given histogram is modified at the same time as you are printing it out (with all those fancy columns of starts, etc), copy one histogram at a time locally, reset it to zero immediately, then emit it from local copy. That way it's safe on output and minimizes the race between reading and resetting to all zeros.

Comment (ethercflow) [reply libbpf-tools/softirqs.c]: @anakryiko Fixed, thanks!

File: libbpf-tools/softirqs.c (hunk 7 above)
Comment (ethercflow) [reply libbpf-tools/softirqs.c]: Fixed, thanks!

File: libbpf-tools/softirqs.c [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+
+static int print_hist(struct softirqs_bpf__bss *bss)
+{
+	const char *units = env.nanoseconds ? "nsecs" : "usecs";
+	__u32 vec;
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++) {
+		struct hist hist = bss->hists[vec];
+
+		memset(&bss->hists[vec], 0, sizeof(bss->hists[vec]));
```
Comment (anakryiko) [inline libbpf-tools/softirqs.c:166]: you copied struct implicitly (through struct assignment), but set it to zero through memset :)) this line could have been:
```
bss->hists[vec] = zero;
```

**Review by anakryiko — APPROVED** [review]
Looks good. Approved.

File: libbpf-tools/softirqs.c (hunk 9 above)
Comment (ethercflow) [reply libbpf-tools/softirqs.c]: Oh,  sorry for this, haha. Fixed, thanks!

**Comment by ethercflow** [comment]
Hi @yonghong-song , PTAL when you have time, thanks! :)

**Comment by yonghong-song** [comment]
[buildbot, test this please]
